// @tryghost/prometheus-metrics/build/PrometheusClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrometheusClient = void 0;
const prom_client_1 = __importDefault(require("prom-client"));
const logging_1 = __importDefault(require("@tryghost/logging"));
const errors_1 = __importDefault(require("@tryghost/errors"));
/**
 * A client for exporting metrics to Prometheus, based on prom-client
 */
class PrometheusClient {
    /**
     * Creates a new PrometheusClient instance
     * @param prometheusConfig - The configuration for the PrometheusClient
     */
    constructor(prometheusConfig = {}) {
        this.config = prometheusConfig;
        this.client = prom_client_1.default;
        this.prefix = 'ghost_';
    }
    client;
    config;
    prefix;
    gateway; // public for testing
    pushInterval;
    /**
     * Initializes the prometheus client, setting up the pushgateway if enabled
     */
    init() {
        this.collectDefaultMetrics();
        if (this.config.pushgateway?.enabled) {
            const gatewayUrl = this.config.pushgateway.url || 'http://localhost:9091';
            const interval = this.config.pushgateway.interval || 5000;
            this.gateway = new prom_client_1.default.Pushgateway(gatewayUrl);
            this.pushInterval = setInterval(() => {
                this.pushMetrics();
            }, interval);
        }
    }
    /**
     * Pushes metrics to the pushgateway, if enabled
     */
    async pushMetrics() {
        if (this.config.pushgateway?.enabled && this.gateway) {
            const jobName = this.config.pushgateway?.jobName || 'ghost';
            try {
                await this.gateway.pushAdd({ jobName });
                logging_1.default.debug('Metrics pushed to pushgateway - jobName: ', jobName);
            }
            catch (err) {
                let error;
                if (typeof err === 'object' && err !== null && 'code' in err) {
                    error = new errors_1.default.InternalServerError({ message: 'Error pushing metrics to pushgateway: ' + err.code, code: err.code });
                }
                else {
                    error = new errors_1.default.InternalServerError({ message: 'Error pushing metrics to pushgateway: Unknown error' });
                }
                logging_1.default.error(error);
            }
        }
    }
    /**
     * Shuts down the prometheus client cleanly
     */
    stop() {
        // Clear the push interval
        if (this.pushInterval) {
            clearInterval(this.pushInterval);
        }
    }
    /**
     * Tells prom-client to collect default metrics
     * Only called once on init
     */
    collectDefaultMetrics() {
        this.client.collectDefaultMetrics({ prefix: this.prefix });
    }
    /**
     * Handles metrics requests to serve the /metrics endpoint
     * @param req - The request object
     * @param res - The response object
     */
    async handleMetricsRequest(req, res) {
        try {
            res.set('Content-Type', this.getContentType());
            res.end(await this.getMetrics());
        }
        catch (err) {
            if (err instanceof Error && err.message) {
                res.status(500).end(err.message);
            }
            else {
                res.status(500).end('Unknown error');
            }
        }
    }
    /**
     * Returns the metrics from the registry
     */
    async getMetrics() {
        return this.client.register.metrics();
    }
    /**
     * Returns the content type for the metrics
     */
    getContentType() {
        return this.client.register.contentType;
    }
}
exports.PrometheusClient = PrometheusClient;
